Soundtouch and sbsms effects in a module

Paul-Licameli opened the pull request:
Resolves #1131

Making the soundtouch/sbsms effects the first example of old features separated into a new module.

This choice is interesting because it also removes all use of two third-party libraries from the Audacity executable.  Only the plug-in depends on them.  To make the link work on Linux as well, the libraries need to compile as position-independent code.

These effects are Change Speed, Change Pitch, Change Tempo, and Sliding Stretch.  Audacity should simply run and you should see those effects as usual in the menus.

Some support code is included, that is needed for modules in general.  HOWEVER -- this may only be complete for macOS.  Linux and Windows builds do not yet put the modules into the correct folder.

SteveDaulton commented:
I've not installed your new version Paul. I just ran it from the artefact using Leland's trick:

`AUDACITY_MODULES_PATH=$(pwd)/lib/audacity/modules ./audacity`

The mod-soundtouch module was loaded correctly.
Both SoundTouch and SBSMS work by default.

If I disable the module, Change Pitch, Change Tempo and Sliding Stretch effects "disappear". Why are they not available with SBSMS only? Why did "Sliding Stretch" disappear when it does not use SoundTouch?

On re-enabling mod-soundtouch, the "missing" effects are still disabled.It is extremely common for new users to "click around" to discover how to use Audacity, so this is a bear trap that is likely to catch a lot of new users. 

IMO, these effects should be "temporarily" disabled when the required libraries are unavailable and should automatically be re-enabled when the libraries are available, (unless the user has explicitly disabled the effects through the plug-in manager).

There should also be a warning when disabling mod-soundtouch to say that doing so will disable those effects.

Paul-Licameli commented:
On Tuesday, May 18, 2021, Steve Daulton ***@***.***> wrote:

> I've not installed your new version Paul. I just ran it from the artefact
> using Leland's trick:
>
> AUDACITY_MODULES_PATH=$(pwd)/lib/audacity/modules ./audacity
>
>
I think something needs to change so the trick is not needed.  What?
Either CMake needs to put the modules in another place or the module
manager needs to look in the place where they are.

> The mod-soundtouch module was loaded correctly.
> Both SoundTouch and SBSMS work by default.
>
> If I disable the module, Change Pitch, Change Tempo and Sliding Stretch
> effects "disappear". Why are they not available with SBSMS only? Why did
> "Sliding Stretch" disappear when it does not use SoundTouch?
>
The module is misnamed maybe.  It contains both sound touch and Sbsms
effects.  Two of them are written with alternative implementations.

> On re-enabling mod-soundtouch, the "missing" effects are still disabled.It
> is extremely common for new users to "click around" to discover how to use
> Audacity, so this is a bear trap that is likely to catch a lot of new users.
>
That is a bug.

> IMO, these effects should be "temporarily" disabled when the required
> libraries are unavailable and should automatically be re-enabled when the
> libraries are available, (unless the user has explicitly disabled the
> effects through the plug-in manager).
>
> There should also be a warning when disabling mod-soundtouch to say that
> doing so will disable those effects.
>
That would add complication.  Should there be a warning associated with
every feature that is eventually sepa

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/audacity/audacity/pull/902#issuecomment-843007130>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZBGYIQ5DBFGMNG3XAOK43TOIWTJANCNFSM45BMTNOQ>
> .
>

Paul-Licameli commented:
Leland, some more commits for review in this version of the branch.  Steve,
please watch for these artifacts:

https://github.com/Paul-Licameli/audacity/actions/runs/854801206

I hope that in this build, the soundtouch module will simply load without
needing the environment variable setup on Linux.

I also expect it will just load (as it didn't before) on Windows.

To fix Linux, I added another path to the audacityPathList computed in
AudacityApp::OnInit().

But to fix Windows, I changed some CMake code.  The Windows packages were
not making a subdirectory called "modules" as was intended, and I think the
explanation is on this page :
https://cmake.org/cmake/help/v3.0/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html
and so the fix required CMake changes.

On Mac there was no mismatch before between the folders that modules went
to, and the folders that were checked for modules by ModuleManager.

PRL


On Tue, May 18, 2021 at 5:57 AM Paul Licameli ***@***.***>
wrote:

>
>
> On Tuesday, May 18, 2021, Steve Daulton ***@***.***> wrote:
>
>> I've not installed your new version Paul. I just ran it from the artefact
>> using Leland's trick:
>>
>> AUDACITY_MODULES_PATH=$(pwd)/lib/audacity/modules ./audacity
>>
>>
> I think something needs to change so the trick is not needed.  What?
> Either CMake needs to put the modules in another place or the module
> manager needs to look in the place where they are.
>
>> The mod-soundtouch module was loaded correctly.
>> Both SoundTouch and SBSMS work by default.
>>
>> If I disable the module, Change Pitch, Change Tempo and Sliding Stretch
>> effects "disappear". Why are they not available with SBSMS only? Why did
>> "Sliding Stretch" disappear when it does not use SoundTouch?
>>
> The module is misnamed maybe.  It contains both sound touch and Sbsms
> effects.  Two of them are written with alternative implementations.
>
>> On re-enabling mod-soundtouch, the "missing" effects are still
>> disabled.It is extremely common for new users to "click around" to discover
>> how to use Audacity, so this is a bear trap that is likely to catch a lot
>> of new users.
>>
> That is a bug.
>
>> IMO, these effects should be "temporarily" disabled when the required
>> libraries are unavailable and should automatically be re-enabled when the
>> libraries are available, (unless the user has explicitly disabled the
>> effects through the plug-in manager).
>>
>> There should also be a warning when disabling mod-soundtouch to say that
>> doing so will disable those effects.
>>
> That would add complication.  Should there be a warning associated with
> every feature that is eventually sepa
>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/audacity/audacity/pull/902#issuecomment-843007130>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACZBGYIQ5DBFGMNG3XAOK43TOIWTJANCNFSM45BMTNOQ>
>> .
>>
>

imciner2 commented on src/AudacityApp.cpp:
```diff
@@ -1149,6 +1149,8 @@ bool AudacityApp::OnInit()
    wxString progPath = wxPathOnly(argv[0]);
    FileNames::AddUniquePathToPathList(progPath, audacityPathList);
+   // Add the path to modules:
+   FileNames::AddUniquePathToPathList(progPath + L"/lib/audacity", audacityPathList);
```
Is this correct for installed versions? In that case Audacity would be `/usr/bin/audacity`, but the modules would be `/usr/lib/audacity`, and I think this code would give `/usr/bin/lib/audacity` if I am reading it correctly.

Paul-Licameli commented:
I believe what I wrote is sufficient to correct the behavior of executables
as they are built for Ubuntu in GitHub Actions.

Whether it is necessary or correct for other installations, I don't know
and will let those more expert in Linux decide.

PRL


On Tue, May 18, 2021 at 7:43 PM Ian McInerney ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/AudacityApp.cpp
> <https://github.com/audacity/audacity/pull/902#discussion_r634821277>:
>
> > @@ -1149,6 +1149,8 @@ bool AudacityApp::OnInit()
>
>     wxString progPath = wxPathOnly(argv[0]);
>     FileNames::AddUniquePathToPathList(progPath, audacityPathList);
> +   // Add the path to modules:
> +   FileNames::AddUniquePathToPathList(progPath + L"/lib/audacity", audacityPathList);
>
> Is this correct for installed versions? In that case Audacity would be
> /usr/bin/audacity, but the modules would be /usr/lib/audacity, and I
> think this code would give /usr/bin/lib/audacity if I am reading it
> correctly.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/audacity/audacity/pull/902#pullrequestreview-662663815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZBGYKPXDYWZYL55UVJSCDTOL3RDANCNFSM45BMTNOQ>
> .
>
